src/array/map/mutable.rs

```diff
@@ -0,0 +1,192 @@
+    /// # Errors
+    /// This function errors if:
+    /// * The `data_type`'s physical type is not [`crate::datatypes::PhysicalType::Map`]
+    /// * The fields' `data_type` is not equal to the inner field of `data_type`
+    pub fn try_new(data_type: DataType, values: Vec<Box<dyn MutableArray>>) -> Result<Self> {
+        let field = MutableStructArray::try_from_data(
+            MapArray::get_field(&data_type).data_type().clone(),
+            values,
+            None,
+        )?;
+        Ok(Self {
+            data_type,
+            offsets: vec![0i32; 1],
+            field,
+            validity: None,
+        })
+    }
+
+    /// Returns the length of this array
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.offsets.len() - 1
+    }
+
+    /// Mutable reference to the field
+    pub fn mut_field(&mut self) -> &mut MutableStructArray {
+        &mut self.field
+    }
+
+    /// Reference to the field
+    pub fn field(&self) -> &MutableStructArray {
+        &self.field
+    }
+
+    /// Get a mutable reference to the keys and values arrays
+    pub fn keys_values<K, V>(&mut self) -> Option<(&mut K, &mut V)>
+    where
+        K: MutableArray + 'static,
+        V: MutableArray + 'static,
+    {
+        let [keys, values]: &mut [_; 2] =
+            self.field.mut_values().as_mut_slice().try_into().unwrap();
+
+        match (
+            keys.as_mut_any().downcast_mut(),
+            values.as_mut_any().downcast_mut(),
+        ) {
+            (Some(keys), Some(values)) => Some((keys, values)),
+            _ => None,
+        }
+    }
+
+    /// Call this once for each "row" of children you push.
+    pub fn push(&mut self, valid: bool) {
+        match &mut self.validity {
+            Some(validity) => validity.push(valid),
+            None => match valid {
+                true => (),
+                false => self.init_validity(),
+            },
+        }
+    }
+
+    #[inline]
+    /// Needs to be called when a valid value was extended to this array.
+    pub fn try_push_valid(&mut self) -> Result<()> {
+        let size = self.field.len();
+        let size = <i32 as Index>::from_usize(size).ok_or(Error::Overflow)?;
+        assert!(size >= *self.offsets.last().unwrap());
+        self.offsets.push(size);
+        if let Some(validity) = &mut self.validity {
+            validity.push(true)
+        }
+        Ok(())
+    }
+
+    fn push_null(&mut self) {
+        self.field.push(false);
+        self.push(false);
+    }
+
+    fn init_validity(&mut self) {
+        let mut validity = MutableBitmap::with_capacity(self.field.len());
+        let len = self.len();
+        if len > 0 {
+            validity.extend_constant(len, true);
+            validity.set(len - 1, false);
+        }
+        self.validity = Some(validity);
+    }
+
+    fn take_into(&mut self) -> MapArray {
+        MapArray::from_data(
+            self.data_type.clone(),
+            std::mem::take(&mut self.offsets).into(),
```

Comment:
if we take here, the offsets end up being empty, which I believe violates an invariant of `self`

Reply:
Ah yeah, the API on `MutableArray` is unclear here. Shouldn't `as_box` and `as_arc` take `self` instead of `&mut self`?

Reply:
Yes, but that would require `MutableArray` to not be object safe, and we need that for using it as a trait object.